### PR TITLE
fuchsia: Add a SoftwareSurfaceProducer for debug

### PR DESCRIPTION
**Pull request by arbreng**
This adds the ability to use software rendering for flutter-on-fuchsia, enabling some desired use-cases (also, it was fun to write)

Tests: Ran flutter-embedder-test2 w/ software forced on, ran workstation
Fixes: https://bugs.fuchsia.dev/p/fuchsia/issues/detail?id=88467

### CODE REVIEW

File: shell/platform/fuchsia/flutter/engine.cc [hunk 1]
```diff
@@ -436,7 +454,7 @@ void Engine::Initialize(
                                ->GetTaskRunner()
                                .get(),
                            shell.GetTaskRunners().GetRasterTaskRunner().get(),
-                           surface_producer_.value(), 1024, 600,
+                           surface_producer_, SkISize::Make(1024, 600),
```
Comment (arbreng) [inline shell/platform/fuchsia/flutter/engine.cc:457]: Drive-by fix

File: shell/platform/fuchsia/flutter/engine.cc [hunk 2]
```diff
@@ -827,19 +848,16 @@ void Engine::WriteProfileToTrace() const {
...
+    std::shared_ptr<SurfaceProducer> surface_producer,
+    SkISize size,
     std::shared_ptr<flutter::AssetManager> asset_manager,
     std::optional<const std::vector<std::string>> skp_names,
     std::optional<std::function<void(uint32_t)>> completion_callback) {
-  SkISize size = SkISize::Make(width, height);
-
   // We use a raw pointer here because we want to keep this alive until all gpu
   // work is done and the callbacks skia takes for this are function pointers
   // so we are unable to use a lambda that captures the smart pointer.
```
Comment (chandarrengoog) [inline shell/platform/fuchsia/flutter/engine.cc:858]: Does this comment need to be updated?

File: shell/platform/fuchsia/flutter/software_surface.cc [hunk 3]
```diff
@@ -0,0 +1,404 @@
...
+  zx_status_t bind_status = sysmem_allocator->BindSharedCollection(
+      std::move(local_token), buffer_collection.NewRequest());
+  if (bind_status != ZX_OK) {
+    FML_LOG(ERROR) << "SoftwareSurface: Failed to bind collection token: "
+                   << zx_status_get_string(bind_status);
+    return false;
+  }
+
+  fuchsia::sysmem::BufferCollectionConstraints constraints;
+  constraints.min_buffer_count = 1;
```
Comment (uysalere) [inline shell/platform/fuchsia/flutter/software_surface.cc:174]: min_buffer_count_for_camping might be set on the Scenic side, and actually there might be multiple buffers allocated for this BufferCollection. If you know that you are going to need multiple of these buffers, there can be an optimization by allocating more than one here and adding to the pool.  Good to leave as a TODO to avoid code complexity here.

File: shell/platform/fuchsia/flutter/software_surface.cc [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+  image_constraints.max_coded_width_times_coded_height =
+      std::numeric_limits<uint32_t>::max();
+  image_constraints.pixel_format.type =
+      fuchsia::sysmem::PixelFormatType::R8G8B8A8;
+  image_constraints.color_spaces_count = 1;
+  image_constraints.color_space[0].type = fuchsia::sysmem::ColorSpaceType::SRGB;
+  image_constraints.pixel_format.has_format_modifier = true;
+  image_constraints.pixel_format.format_modifier.value =
+      fuchsia::sysmem::FORMAT_MODIFIER_LINEAR;
+  image_constraints.min_bytes_per_row = static_cast<uint32_t>(size.fWidth) * 4;
```
Comment (uysalere) [inline shell/platform/fuchsia/flutter/software_surface.cc:205]: These should be implied when you FORMAT_MODIFIER_LINEAR, but no harm in specifying them again.

File: shell/platform/fuchsia/flutter/software_surface.cc [hunk 5]
```diff
@@ -0,0 +1,404 @@
...
+    FML_LOG(ERROR) << "SoftwareSurface: Failed to allocate: "
+                   << zx_status_get_string(allocation_status);
+    return false;
+  }
+  if (wait_for_allocated_status != ZX_OK) {
+    FML_LOG(ERROR) << "SoftwareSurface: Failed to wait for allocate: "
+                   << zx_status_get_string(wait_for_allocated_status);
+    return false;
+  }
+  if (buffer_collection_info.settings.image_format_constraints.pixel_format
```
Comment (uysalere) [inline shell/platform/fuchsia/flutter/software_surface.cc:230]: I dont think these checks are necessary. We can trust sysmem API to fail at WaitForBuffersAllocated() if our constraints arent applied.

File: shell/platform/fuchsia/flutter/software_surface.cc [hunk 6]
```diff
@@ -0,0 +1,404 @@
...
+  fuchsia::sysmem::BufferCollectionConstraints constraints;
+  constraints.min_buffer_count = 1;
+  constraints.usage.cpu =
+      fuchsia::sysmem::cpuUsageWrite | fuchsia::sysmem::cpuUsageWriteOften;
+  constraints.has_buffer_memory_constraints = true;
+  constraints.buffer_memory_constraints.min_size_bytes = 0;
+  constraints.buffer_memory_constraints.max_size_bytes = 0xffffffff;
+  constraints.buffer_memory_constraints.physically_contiguous_required = false;
+  constraints.buffer_memory_constraints.secure_required = false;
+  constraints.buffer_memory_constraints.ram_domain_supported = true;
```
Comment (uysalere) [inline shell/platform/fuchsia/flutter/software_surface.cc:182]: If you allow RAM domain, you need to flush after writing to it. See https://cs.opensource.google/fuchsia/fuchsia/+/master:src/ui/scenic/lib/gfx/tests/scenic_pixeltest.cc;drc=5796f5829de3eb33efc3c51dd739eded4f9abf9b;l=1408

Comment (jbauman42) [reply shell/platform/fuchsia/flutter/software_surface.cc:182]: The ARM Mali GPU driver (used by scenic on some ARM platforms) requires RAM domain. If you don't flush after writing, you can get tearing and corruption when reading from the GPU. If you read from the buffer and another client wrote to it you also need to invalidate before the read, but that doesn't seem to be necessary here.

**Review by chandarrengoog — COMMENTED** [review]
It seems that this PR has issues building: https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8830919180832446545/+/u/build_fuchsia_debug_x64_flutter_shell_platform_fuchsia:fuchsia_fuchsia_tests/stdout#:~:text=1%20error%20generated.

Generally LG; deferring approval to Emircan since I'm not the subject matter expert. 🙂

File: shell/platform/fuchsia/flutter/software_surface.cc [hunk 7]
```diff
@@ -0,0 +1,404 @@
...
+    callback();
+  }
+}
+
+void SoftwareSurface::OnHandleReady(async_dispatcher_t* dispatcher,
+                                    async::WaitBase* wait,
+                                    zx_status_t status,
+                                    const zx_packet_signal_t* signal) {
+  if (status != ZX_OK) {
+    << zx_status_get_string(status);
```
Comment (chandarrengoog) [inline shell/platform/fuchsia/flutter/software_surface.cc:397]: Maybe `FML_LOG(ERROR) << zx_status_get_string(status);`?

File: shell/platform/fuchsia/flutter/software_surface_producer.cc [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+  }
+}
+
+std::unique_ptr<SoftwareSurface> SoftwareSurfaceProducer::CreateSurface(
+    const SkISize& size) {
+  TRACE_EVENT2("flutter", "SoftwareSurfacePool::CreateSurface", "width",
+               size.width(), "height", size.height());
+  auto surface =
+      std::make_unique<SoftwareSurface>(sysmem_allocator_, flatland_allocator_,
+                                        scenic_session_, size, buffer_id_++);
```
Comment (chandarrengoog) [inline shell/platform/fuchsia/flutter/software_surface_producer.cc:149]: Should we use a variant here between flatland and scenic?

File: shell/platform/fuchsia/flutter/software_surface.cc [hunk 9]
```diff
@@ -0,0 +1,404 @@
...
+}  // namespace
+
+SoftwareSurface::SoftwareSurface(
+    fuchsia::sysmem::AllocatorSyncPtr& sysmem_allocator,
+    fuchsia::ui::composition::AllocatorPtr& flatland_allocator,
+    scenic::Session* session,
+    const SkISize& size,
+    uint32_t buffer_id)
+    : session_(session), wait_(this) {
+  FML_CHECK(session_ || flatland_allocator.is_bound());
```
Comment (chandarrengoog) [inline shell/platform/fuchsia/flutter/software_surface.cc:37]: Should we do `  FML_CHECK((session_ == nullptr) ^ flatland_allocator.is_bound());` instead?

Having both seems like an invalid state.

File: shell/platform/fuchsia/flutter/software_surface_producer.cc [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+  if (exact_match_it != available_surfaces_.end()) {
+    TRACE_EVENT_INSTANT0("flutter", "Exact match found");
+    surface = std::move(*exact_match_it);
+    available_surfaces_.erase(exact_match_it);
+  } else {
+    surface = CreateSurface(size);
+  }
+
+  if (surface == nullptr) {
+    FML_LOG(ERROR) << "SoftwareSurfaceProducer: Could not acquire surface";
```
Comment (akbiggs) [inline shell/platform/fuchsia/flutter/software_surface_producer.cc:90]: nit: Do we need the class name in these logs given that software_surface_producer.cc should already be tagged on the logs?

File: shell/platform/fuchsia/flutter/software_surface_producer.cc [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+    FML_LOG(ERROR) << "SoftwareSurfaceProducer: Created surface is invalid";
+    return nullptr;
+  }
+  trace_surfaces_created_++;
+  return surface;
+}
+
+void SoftwareSurfaceProducer::RecycleSurface(
+    std::unique_ptr<SoftwareSurface> surface) {
+  // The surface may have become invalid (for example it the fences could
```
Comment (akbiggs) [inline shell/platform/fuchsia/flutter/software_surface_producer.cc:160]: nit: "for example if"

File: shell/platform/fuchsia/flutter/software_surface_producer.cc [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+  // not be reset).
+  if (!surface->IsValid()) {
+    return;
```
Comment (akbiggs) [inline shell/platform/fuchsia/flutter/software_surface_producer.cc:163]: Even if the root cause was elsewhere, should we log an error here anyways?

File: shell/platform/fuchsia/flutter/software_surface_producer.cc [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+  }
+  TraceStats();
+}
+
+void SoftwareSurfaceProducer::RecyclePendingSurface(uintptr_t surface_key) {
+  // Before we do anything, we must clear the surface from the collection of
+  // pending surfaces.
+  auto found_in_pending = pending_surfaces_.find(surface_key);
+  if (found_in_pending == pending_surfaces_.end()) {
+    return;
```
Comment (akbiggs) [inline shell/platform/fuchsia/flutter/software_surface_producer.cc:182]: Should we log an error here?

**Review by akbiggs — APPROVED** [review]
Looks good, comments are all nits.

File: shell/platform/fuchsia/flutter/software_surface.h [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+  async::WaitMethod<SoftwareSurface, &SoftwareSurface::OnHandleReady> wait_;
+  std::function<void()> pending_on_writes_committed_;
+  ReleaseImageCallback release_image_callback_;
+
+  fuchsia::ui::composition::BufferCollectionImportToken import_token_;
+  zx::event acquire_event_;
+  zx::event release_event_;
+  uint32_t buffer_id_ = 0;
+
+  size_t age_ = 0;
```
Comment (akbiggs) [inline shell/platform/fuchsia/flutter/software_surface.h:109]: Add a docstring for `age_` explaining what it represents

File: shell/platform/fuchsia/flutter/software_surface.cc [hunk 15]
```diff
@@ -0,0 +1,404 @@
...
+    fuchsia::ui::composition::AllocatorPtr& flatland_allocator,
+    const SkISize& size,
+    uint32_t buffer_id) {
+  if (size.isEmpty()) {
+    FML_LOG(ERROR)
+        << "SoftwareSurface: Failed to allocate surface, size is empty";
+    return false;
+  }
+
+  fuchsia::sysmem::BufferCollectionTokenSyncPtr local_token;
```
Comment (akbiggs) [inline shell/platform/fuchsia/flutter/software_surface.cc:111]: Would it be useful to leave a comment here explaining the roles of:

- local_token
- scenic_token
- import_token
- export_token?

It's hard for me to tell the differences between them.

File: shell/platform/fuchsia/flutter/software_surface.cc [hunk 16]
```diff
@@ -0,0 +1,404 @@
...
+  zx_status_t allocate_status =
+      sysmem_allocator->AllocateSharedCollection(local_token.NewRequest());
+  if (allocate_status != ZX_OK) {
+    FML_LOG(ERROR) << "SoftwareSurface: Failed to allocate collection: "
+                   << zx_status_get_string(allocate_status);
+    return false;
+  }
+  fuchsia::sysmem::BufferCollectionTokenSyncPtr scenic_token;
+  zx_status_t duplicate_status = local_token->Duplicate(
```
Comment (akbiggs) [inline shell/platform/fuchsia/flutter/software_surface.cc:120]: Would DuplicateSync also work here instead of separately Duplicate and Sync calls? https://cs.opensource.google/fuchsia/fuchsia/+/main:sdk/fidl/fuchsia.sysmem/collection.fidl;l=36?q=%22DuplicateSync%22%20f:fidl$&ss=fuchsia

File: shell/platform/fuchsia/flutter/software_surface.h [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+  void PushSessionImageSetupOps(scenic::Session* session);
+
+  void Reset();
+
+  scenic::Session* session_;
+  uint32_t image_id_ = 0;
+
+  sk_sp<SkSurface> sk_surface_;
+  async::WaitMethod<SoftwareSurface, &SoftwareSurface::OnHandleReady> wait_;
+  std::function<void()> pending_on_writes_committed_;
```
Comment (akbiggs) [inline shell/platform/fuchsia/flutter/software_surface.h:101]: Add docstring explaining why we queue this and when it should get flushed.

File: shell/platform/fuchsia/flutter/software_surface.h [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+      uint32_t buffer_id);
+
+  bool CreateFences();
+
+  void PushSessionImageSetupOps(scenic::Session* session);
+
+  void Reset();
+
+  scenic::Session* session_;
+  uint32_t image_id_ = 0;
```
Comment (akbiggs) [inline shell/platform/fuchsia/flutter/software_surface.h:97]: Should we add a docstring explaining what "space" this ID is in, i.e. where it needs to come from/be used?

File: shell/platform/fuchsia/flutter/software_surface.h [hunk 19]
```diff
@@ -0,0 +1,115 @@
...
+
+  sk_sp<SkSurface> sk_surface_;
+  async::WaitMethod<SoftwareSurface, &SoftwareSurface::OnHandleReady> wait_;
+  std::function<void()> pending_on_writes_committed_;
+  ReleaseImageCallback release_image_callback_;
+
+  fuchsia::ui::composition::BufferCollectionImportToken import_token_;
+  zx::event acquire_event_;
+  zx::event release_event_;
+  uint32_t buffer_id_ = 0;
```
Comment (akbiggs) [inline shell/platform/fuchsia/flutter/software_surface.h:107]: Similarly should we add a comment here explaining what "space" this ID is in?

File: shell/platform/fuchsia/flutter/software_surface_producer.cc (hunk 8 above)
Comment (akbiggs) [reply shell/platform/fuchsia/flutter/software_surface_producer.cc:149]: +1, I feel like this would make the code easier to follow even if it's fairly redundant between the two. The amount of FML_CHECKs/functions that don't do anything based on the presence of a session make me feel like the class will be difficult to reason about.

**Review by akbiggs — COMMENTED** [review]
Sorry, didn't see the SoftwareSurface code in the review somehow. Still reading through that.

**Comment by chinmaygarde** [comment]
This patch has an LGTM but it looks like the review is still in progress. Is this stalled?

**Comment by arbreng** [comment]
@chinmaygarde yeah, I got yanked onto higher-priority items so didn't get a chance to address feedback.   Going to do that this morning and land it.

File: shell/platform/fuchsia/flutter/software_surface.h (hunk 18 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface.h:97]: Even better, this can be made a scenic::ResourceId which makes the type/space obvious.

File: shell/platform/fuchsia/flutter/software_surface.cc (hunk 15 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface.cc:111]: Done

File: shell/platform/fuchsia/flutter/software_surface.cc (hunk 16 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface.cc:120]: Done

File: shell/platform/fuchsia/flutter/software_surface.h (hunk 17 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface.h:101]: Renamed to be more descriptive and added comments

File: shell/platform/fuchsia/flutter/software_surface.h (hunk 19 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface.h:107]: Done, moved the counter to a static in the same class.  It's a lot clearer now

File: shell/platform/fuchsia/flutter/software_surface_producer.cc (hunk 8 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface_producer.cc:149]: I agree that it makes the code much harder to reason about.

The thing is - The scenic dependency here is a false dependency; this code doesn't actually *need* a dependency on scenic and it is only here due to smelly legacy code.

See https://bugs.fuchsia.dev/p/fuchsia/issues/detail?id=88466 for a fuller description of what is going on and how to fix this.

Since a variant would encourage deeper intertwining of the code, and the goal is to break the dependency, I don't want to use one here.  Instead, we can reduce the code complexity by fixing 88466.

File: shell/platform/fuchsia/flutter/software_surface.cc [hunk 20]
```diff
@@ -0,0 +1,404 @@
...
+      std::numeric_limits<uint32_t>::max();
+  image_constraints.pixel_format.type =
+      fuchsia::sysmem::PixelFormatType::R8G8B8A8;
+  image_constraints.color_spaces_count = 1;
+  image_constraints.color_space[0].type = fuchsia::sysmem::ColorSpaceType::SRGB;
+  image_constraints.pixel_format.has_format_modifier = true;
+  image_constraints.pixel_format.format_modifier.value =
+      fuchsia::sysmem::FORMAT_MODIFIER_LINEAR;
+  image_constraints.min_bytes_per_row = static_cast<uint32_t>(size.fWidth) * 4;
+  image_constraints.max_bytes_per_row = static_cast<uint32_t>(size.fWidth) * 4;
```
Comment (jbauman42) [inline shell/platform/fuchsia/flutter/software_surface.cc:206]: Don't set max_bytes_per_row, since that may make it impossible to allocate in some circumstances.

File: shell/platform/fuchsia/flutter/software_surface.cc [hunk 21]
```diff
@@ -0,0 +1,404 @@
...
+        << "SoftwareSurface: Failed to allocate; buffer isn't present.";
+    return false;
+  }
+
+  const zx::vmo& image_vmo = buffer_collection_info.buffers[0].vmo;
+  const uint32_t image_vmo_bytes =
+      buffer_collection_info.settings.buffer_settings.size_bytes;
+  const uint64_t vmo_offset =
+      buffer_collection_info.buffers[0].vmo_usable_start;
+  const size_t vmo_stride = buffer_collection_info.settings
```
Comment (jbauman42) [inline shell/platform/fuchsia/flutter/software_surface.cc:251]: The stride should be calculated using the formula at https://osscs.corp.google.com/fuchsia/fuchsia/+/main:zircon/system/ulib/image-format/image_format.cc;l=1142?q=ImageformatM&ss=fuchsia

File: shell/platform/fuchsia/flutter/software_surface.cc (hunk 7 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface.cc:397]: Oops, fixed

File: shell/platform/fuchsia/flutter/software_surface.cc (hunk 9 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface.cc:37]: Done

File: shell/platform/fuchsia/flutter/software_surface.cc [hunk 22]
```diff
@@ -0,0 +1,404 @@
...
+  constraints.image_format_constraints_count = 1;
+  fuchsia::sysmem::ImageFormatConstraints& image_constraints =
+      constraints.image_format_constraints[0];
+  image_constraints = fuchsia::sysmem::ImageFormatConstraints();
+  image_constraints.min_coded_width = static_cast<uint32_t>(size.fWidth);
+  image_constraints.min_coded_height = static_cast<uint32_t>(size.fHeight);
+  image_constraints.max_coded_width = static_cast<uint32_t>(size.fWidth);
+  image_constraints.max_coded_height = static_cast<uint32_t>(size.fHeight);
+  image_constraints.min_bytes_per_row = 0;
+  image_constraints.max_bytes_per_row = std::numeric_limits<uint32_t>::max();
```
Comment (jbauman42) [inline shell/platform/fuchsia/flutter/software_surface.cc:195]: Set to 0 to use the default value.

File: shell/platform/fuchsia/flutter/software_surface.cc [hunk 23]
```diff
@@ -0,0 +1,404 @@
...
+  image_constraints.max_coded_width_times_coded_height =
```
Comment (jbauman42) [inline shell/platform/fuchsia/flutter/software_surface.cc:196]: Set to 0 to use the default value.

File: shell/platform/fuchsia/flutter/software_surface.cc [hunk 24]
```diff
@@ -0,0 +1,404 @@
...
+    return false;
+  }
+
+  fuchsia::sysmem::BufferCollectionConstraints constraints;
+  constraints.min_buffer_count = 1;
+  constraints.usage.cpu =
+      fuchsia::sysmem::cpuUsageWrite | fuchsia::sysmem::cpuUsageWriteOften;
+  constraints.has_buffer_memory_constraints = true;
+  constraints.buffer_memory_constraints.min_size_bytes = 0;
+  constraints.buffer_memory_constraints.max_size_bytes = 0xffffffff;
```
Comment (jbauman42) [inline shell/platform/fuchsia/flutter/software_surface.cc:179]: Set to 0 to use the default value.

File: shell/platform/fuchsia/flutter/engine.cc [hunk 25]
```diff
@@ -895,34 +913,42 @@ void Engine::WarmupSkps(
...
+            flush_info.fFinishedProc = [](void* skp_warmup_surface) {
+              delete static_cast<SurfaceProducerSurface*>(skp_warmup_surface);
+            };
+
+            surface_producer->gr_context()->flush(flush_info);
+            surface_producer->gr_context()->submit();
+          }
+        } else {
+          if (i == count - 1) {
+            delete static_cast<SurfaceProducerSurface*>(skp_warmup_surface);
```
[the remark this answers is not in the document]
Comment (arbreng) [reply shell/platform/fuchsia/flutter/engine.cc:949]: Done

File: shell/platform/fuchsia/flutter/software_surface_producer.cc (hunk 12 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface_producer.cc:163]: Done

File: shell/platform/fuchsia/flutter/software_surface_producer.cc (hunk 11 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface_producer.cc:160]: Done

File: shell/platform/fuchsia/flutter/software_surface_producer.cc (hunk 10 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface_producer.cc:90]: Done

File: shell/platform/fuchsia/flutter/software_surface_producer.cc [hunk 26]
```diff
@@ -0,0 +1,236 @@
...
+  char name[ZX_MAX_NAME_LEN];
+  zx_status_t status =
+      zx::process::self()->get_property(ZX_PROP_NAME, name, sizeof(name));
+  return status == ZX_OK ? std::string(name) : std::string();
+}
+
+zx_koid_t GetCurrentProcessId() {
+  zx_info_handle_basic_t info;
+  zx_status_t status = zx::process::self()->get_info(
+      ZX_INFO_HANDLE_BASIC, &info, sizeof(info), nullptr, nullptr);
```
[the remark this answers is not in the document]
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface_producer.cc:34]: Done

File: shell/platform/fuchsia/flutter/software_surface.h [hunk 27]
```diff
@@ -0,0 +1,115 @@
...
+ public:
+  SoftwareSurface(fuchsia::sysmem::AllocatorSyncPtr& sysmem_allocator,
+                  fuchsia::ui::composition::AllocatorPtr& flatland_allocator,
+                  scenic::Session* session,
+                  const SkISize& size,
+                  uint32_t buffer_id);
+
+  ~SoftwareSurface() override;
+
+  size_t GetAllocationSize() const { return 0; }
```
[the remark this answers is not in the document]
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface.h:36]: Done

File: shell/platform/fuchsia/flutter/engine.cc (hunk 2 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/engine.cc:858]: No it's still true

File: shell/platform/fuchsia/flutter/flutter_runner_product_configuration.h [hunk 28]
```diff
@@ -5,18 +5,23 @@
...
+
 #include "flutter/fml/time/time_delta.h"
 
 namespace flutter_runner {
 
 class FlutterRunnerProductConfiguration {
  public:
   FlutterRunnerProductConfiguration() {}
-  FlutterRunnerProductConfiguration(std::string path);
+  explicit FlutterRunnerProductConfiguration(bool software_rendering)
```
[the remark this answers is not in the document]
Comment (arbreng) [reply shell/platform/fuchsia/flutter/flutter_runner_product_configuration.h:17]: I had that in for my own testing purposes; removed it

File: shell/platform/fuchsia/flutter/software_surface_producer.cc [hunk 29]
```diff
@@ -0,0 +1,236 @@
...
+
+namespace flutter_runner {
+
+namespace {
+
+std::string GetCurrentProcessName() {
+  char name[ZX_MAX_NAME_LEN];
+  zx_status_t status =
+      zx::process::self()->get_property(ZX_PROP_NAME, name, sizeof(name));
+  return status == ZX_OK ? std::string(name) : std::string();
```
[the remark this answers is not in the document]
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface_producer.cc:28]: Done

File: shell/platform/fuchsia/flutter/software_surface_producer.cc (hunk 13 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface_producer.cc:182]: Done

File: shell/platform/fuchsia/flutter/software_surface.cc (hunk 20 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface.cc:206]: Done

File: shell/platform/fuchsia/flutter/software_surface.cc (hunk 22 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface.cc:195]: Done

File: shell/platform/fuchsia/flutter/software_surface.cc (hunk 24 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface.cc:179]: Done

File: shell/platform/fuchsia/flutter/software_surface.cc (hunk 23 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface.cc:196]: Done

File: shell/platform/fuchsia/flutter/software_surface.cc (hunk 21 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface.cc:251]: Done

File: shell/platform/fuchsia/flutter/software_surface.cc (hunk 6 above)
Comment (arbreng) [reply shell/platform/fuchsia/flutter/software_surface.cc:199]: Done

**Comment by arbreng** [comment]
All feedback addressed; gonna run the test + workstation one more time to be sure, then land

**Comment by akbiggs** [comment]
FYI the failure here is a flake: https://github.com/flutter/flutter/issues/94492
